nexus/src/app/sagas/instance_create.rs

```diff
@@ -141,16 +144,40 @@ fn saga_instance_create() -> SagaTemplate<SagaInstanceCreate> {
         new_action_noop_undo(sic_create_network_interfaces),
     );
 
-    // Grab an external IP address and port range for the guest's Internet
-    // Gateway, allowing external connectivity.
+    // Allocate an external IP address for the default outbound connectivity,
+    // i.e., instance source NAT. We first allocate the ID, for the same reason
+    // we do so in the case of network interfaces.
     template_builder.append(
-        "external_ip",
-        "ExternalIp",
+        "snat_ip_id",
+        "SnatIpId",
         ActionFunc::new_action(
-            sic_allocate_external_ip,
-            sic_allocate_external_ip_undo,
+            sic_allocate_instance_snat_ip_id,
+            sic_allocate_temporary_external_ips_undo,
```

Comment:
Still working my way through your comments (thanks!) but I think most of them are related to this. The approach here, which I may have botched, was intended to be similar to NICs. We may create multiple records during the IP allocation forward action. In particular, we may successfully create N addresses, and then fail to create address N + 1. But if an action fails, it's undo action is not taken (since it supposedly can only be undone after completing successfully).

That means the action that undoes the (potentially-partial) creation of a bunch of addresses has to be attached to the action _before_ that forward action. So I have an action that creates all the UUIDs, whose undo action is deleting all records with those IDs. The actual creation of the records has no undo action.

James had an alternate strategy here, which also works. Instead of doing multiple fallible operations in one saga node, he adds a fixed set of saga nodes for each possible disk that can be created. Since that's limited by a maximum of 8 disks, there are 8 such nodes, each of which is self-contained and includes the "right" or "expected" undo action. The downside is that there may not be 8 disks, so the saga has some number of actions and undo actions that are no-ops.

Both of these work I think, and we added them around the same time. The solution I arrived at feels more natural to me, but if it's confusing I am happy to move this (and the NIC creation part) to mirror James's approach (in a separate PR).

Reply:
> Still working my way through your comments (thanks!) but I think most of them are related to this. The approach here, which I may have botched, was intended to be similar to NICs. We may create multiple records during the IP allocation forward action. In particular, we may successfully create N addresses, and then fail to create address N + 1. But if an action fails, it's undo action is not taken (since it supposedly can only be undone after completing successfully).
> 
> That means the action that undoes the (potentially-partial) creation of a bunch of addresses has to be attached to the action _before_ that forward action. So I have an action that creates all the UUIDs, whose undo action is deleting all records with those IDs. The actual creation of the records has no undo action.

Ahhhh okay, this makes sense. I still think it's fairly confusing to read this style in the current saga template format. Having the "undo" action be attached to a (seemingly) unrelated node seems also somewhat risky from a refactoring point-of-view. @davepacheco can chime in if he disagrees, but I *think* the intent was that the "undo action should undo the result of this action if it completes successfully, and each action should be atomic".

For example: `sic_allocate_instance_snat_ip` makes a request to the datastore that should "fully complete, or do nothing". So if we attached the undo action (`sic_allocate_temporary_external_ips_undo`) to the same node performing the operation - the `SnatIp` node below - it'll only be invoked if the atomic allocation completed successfully.

Is `sic_allocate_instance_ephemeral_ips` the problem case here, because it is not atomic?
 
> James had an alternate strategy here, which also works. Instead of doing multiple fallible operations in one saga node, he adds a fixed set of saga nodes for each possible disk that can be created. Since that's limited by a maximum of 8 disks, there are 8 such nodes, each of which is self-contained and includes the "right" or "expected" undo action. The downside is that there may not be 8 disks, so the saga has some number of actions and undo actions that are no-ops.
> 
> Both of these work I think, and we added them around the same time. The solution I arrived at feels more natural to me, but if it's confusing I am happy to move this (and the NIC creation part) to mirror James's approach (in a separate PR).

I *think* the approach taken there is preferable because of the aforementioned action/undo action relationship, but I'm also hoping it gets simplified when subsagas are integrated into steno: https://github.com/oxidecomputer/steno/pull/29

Reply:
I agree with the points about confusion and refactoring. It does require that the number of nodes be known statically, though. We do here, but that's definitely a choice I'm making for expediency, not because of any intrinsic or policy limitation I'm aware of.

But in any case, I'd prefer to move this to a separate issue tracking that work, and otherwise integrate this PR (unless there's just a bug!). Are you cool with that?

Reply:
Yeah, I am okay with that (as long as we file an issue to track).

I think I mostly wanted to raise awareness of "there are drawbacks to separating the action from the undo" - and if abiding by that policy is painful today, we should figure out how to make it less painful.

Reply:
Will make an issue prior to merging. Thanks for flagging it!

Reply:
https://github.com/oxidecomputer/omicron/issues/1466